openstack_dashboard/local/dashboards/admin_nci/hvlist/views.py

```diff
@@ -0,0 +1,160 @@
+#    under the License.
+#
+
+from horizon import views, messages
+from openstack_dashboard import api
+from iso8601 import parse_date
+from colorsys import hsv_to_rgb
+import re
+from novaclient.exceptions import NotFound, ClientException
+from django.http import HttpResponseServerError
+
+short_name_p = re.compile(r'^tc(?P<n>\d+)$')
+def short_name(hostname):
+    m = short_name_p.match(hostname)
+    if m:
+        return m.group('n')
+    return hostname
+
+def binary_prefix_scale(b):
+    """Return (prefix, scale) so that (b) B = (b*scale) (prefix)B."""
+    binary = ['', 'Ki', 'Mi', 'Gi', 'Ti', 'Pi', 'Ei']
+    scale = 1
+    for p in binary:
+        if b < 1024:
+            return (p, scale)
+        scale /= 1024.
+        b /= 1024.
+    return (p, scale)
+
+def usage_string(now, tot):
+    """Return something like "0.5 / 1.0 kB" from now,tot=512,1024."""
+    prefix, scale = binary_prefix_scale(tot)
+    now *= scale
+    tot *= scale
+
+    def pretty(n):
+        """Normally it is fine to round display values to nearest int,
+        but for positive values < 1 it is helpful to show that they are nonzero."""
+        if n > 1: return '{:.0f}'.format(n)
+        if n > 0: return '{:.1f}'.format(n)
+        return '0'
+    return '{n} / {t} {p}B'.format(n=pretty(now), t=pretty(tot), p=prefix)
+
+class IndexView(views.APIView):
+    template_name = 'admin/hvlist/index.html'
+
+    def get_data(self, request, context, *args, **kwargs):
+        # grab all the data
+        hypervisors = api.nova.hypervisor_list(request)
+        instances, _ = api.nova.server_list(request, all_tenants=True)
+        projects, _ = api.keystone.tenant_list(request)
+        flavs = api.nova.flavor_list(request)
+
+        # reorganise some
+        projects = {p.id : p for p in projects}
+        flavs = {f.id : f for f in flavs}
+        hypervisor_instances = {} # OS-EXT-SRV-ATTR:host : [instance]
+        for i in instances:
+            try:
+                host = getattr(i, 'OS-EXT-SRV-ATTR:host')
+            except AttributeError:
+                # this will make a pretty box for the error message, although it assumes it's an HTTP error so writes "HTTP None"
+                raise ClientException(None, 'could not get OS-EXT-SRV-ATTR:host attribute of instance')
+            if host not in hypervisor_instances: hypervisor_instances[host] = []
+            if i.flavor['id'] not in flavs:
+                # api.nova.flavor_list (which wraps novaclient.flavors.list) does not get all flavors
+                # so if we have a reference to one that hasn't been retrieved, try looking it up specifically
+                # (wrap this rather trivially in a try block to make the error less cryptic)
+                try:
+                    flavs[i.flavor['id']] = api.nova.flavor_get(request, i.flavor['id'])
+                    messages.warning(request, 'Extra lookup for flavor "'+str(i.flavor['id'])+'"')
```

Comment:
I think maybe a warning pop-up for this case might be a bit annoying.  You've still got the info just via another route so simply continue silently (or use "LOG.debug")?